File: PD Project/Firmware/PD_VRMT_v1_0_6_20240814/main/UART.cpp
#include <Arduino.h>
#include "setup.h"
#include "UART.h"

void(* resetFunc) (void) = 0; //declare reset function at address 0

UART::UART(HX711* weight_fill, HX711* weight_drain, A4950* motor, System* sys, Memory* mem, Sound* buz) {
    _ptr_weight_fill = weight_fill;
    _ptr_weight_drain = weight_drain;
    _ptr_motor = motor;
    _ptr_sys = sys;
    _ptr_mem = mem;
    _ptr_buz = buz;
}

void UART::init(int baud) {
    Serial.begin(BAUDRATE);
    Serial.println("Begin...");
    delay(1000);
    pinMode(PIN_LED, OUTPUT);
}

void UART::check_connection(int pin_connect) {
    bool currently_connected = digitalRead(pin_connect);
    if (currently_connected != _is_connected) {
        if (!currently_connected) {
            // when bluetooth connection lost...
            _is_streaming = false;
            _ptr_motor->set_state(_ptr_motor->STOP);
            digitalWrite(PIN_LED, LOW);
        }
    }
    _is_connected = currently_connected;
}

void UART::data_buffering(byte val) {
    _chksum_tx += val;
    Serial.write(byte(val));
}

void UART::data_buffering(float val) {
    _buffered.floatpt = val;
    for (int i = 0; i < 4; i++) {
        data_buffering(_buffered.binary[i]);
    }
}

void UART::data_send() {
    if (_is_streaming) {
        data_buffering(byte(BLE_HEADER));
        data_buffering(byte(BLE_HEADER));

        _chksum_tx = 0;
        data_buffering(byte(BLE_TX_DATA_LEN));
        data_buffering(byte(_idx));
        data_buffering(_ptr_weight_fill->get_weight());
        data_buffering(_ptr_weight_drain->get_weight());
        data_buffering(_ptr_sys->get_voltage());
        data_buffering(_ptr_sys->get_charging());
        data_buffering(byte(~_chksum_tx));
        
        _idx = (_idx < 99)? _idx + 1 : 0;
    }
}

void UART::data_receive() {
    if (Serial.available()) {
        // read and check header segment
        buffer_shift(_buffer_hd);
        Serial.readBytes(_buffer_hd, 1);

        if ((_buffer_hd[3] == BLE_HEADER) && (_buffer_hd[2] == BLE_HEADER)) {
            _cmd = _buffer_hd[1];
            _len = _buffer_hd[0];
            _chksum_rx = _cmd + _len; 
            
            // compute the checksum
            Serial.readBytes(_buffer_data, _len);
            for (int i = 0; i < _len - 1; i++) {
                _chksum_rx += _buffer_data[i];
            }

            // validate the checksum
            if (_buffer_data[_len - 1] == byte(~_chksum_rx)) {

                // tasks to be done
                switch (_cmd) {

                    case 'R':       // system reset
                        resetFunc();
                        break;

                    case 'D':       // show system data
                        if (!_is_streaming) {
                            Serial.print("Version: "); Serial.println(VERSION);
                            Serial.print("Battery (V): "); Serial.println(_ptr_sys->get_voltage());
                            Serial.print("Charging: "); Serial.println(_ptr_sys->get_charging());
                            Serial.print("Load Cell Fill: "); Serial.print(_ptr_weight_fill->get_zero());
                            Serial.print(" / "); Serial.println(_ptr_weight_fill->get_cal());
                            Serial.print("Load Cell Drain: "); Serial.print(_ptr_weight_drain->get_zero());
                            Serial.print(" / "); Serial.println(_ptr_weight_drain->get_cal());
                        }
                        break;

                    case 'L':       // camera LED on
                        digitalWrite(PIN_LED, HIGH);
                        break;
                    case 'l':       // camera LED off
                        digitalWrite(PIN_LED, LOW);
                        break;

                    case 'S':       // stream start
                        _is_streaming = true;
                        _ptr_motor->set_ready(false);
                        break;
                    case 's':       // stream stop
                        _is_streaming = false;
                        _ptr_motor->set_ready(true);
                        break;

                    case 'Z':       // set zero for load cell 0
                        _ptr_weight_fill->set_zero();
                        _ptr_mem->save();
                        break;
                    case 'z':       // set zero for load cell 1
                        _ptr_weight_drain->set_zero();
                        _ptr_mem->save();
                        break;
                        
                    case 'C':       // calibrate 1kg weight for load cell 0
                        _ptr_weight_fill->set_cal();
                        _ptr_mem->save();
                        break;
                    case 'c':       // calibrate 1kg weight for load cell 1
                        _ptr_weight_drain->set_cal();
                        _ptr_mem->save();
                        break;

                    case 'M':       // motor going up
                        _ptr_motor->set_state(_ptr_motor->UP);
                        break;
                    case 'm':       // motor going down
                        _ptr_motor->set_state(_ptr_motor->DOWN);
                        break;
                    case 'x':       // motor stop
                        _ptr_motor->set_state(_ptr_motor->STOP);
                        break;

                    case 'T':
                        // test motor
                        digitalWrite(PIN_BUZZ, HIGH);
                        _ptr_motor->set_state(_ptr_motor->UP);
                        delay(1000);
                        _ptr_motor->set_state(_ptr_motor->DOWN);
                        delay(1000);
                        _ptr_motor->set_state(_ptr_motor->STOP);
                        digitalWrite(PIN_BUZZ, LOW);
                        break;

                    case 'A':
                        // alarm
                        tone(PIN_BUZZ, 330, 500);
                        break;

                    case 'P':
                        // play music
                        _ptr_buz->beep(5);
                        break;

                    default:
                        break;
                }

                // check if there are more unread messages
                data_receive();
                return true;
            }
        }
        return data_receive();
    }
    return false;
}

void UART::buffer_shift(byte buffer[BLE_HEADER_LEN]) {
    for (int i = BLE_HEADER_LEN - 1; i > 0; i--) {
        buffer[i] = buffer[i-1];
    }
}

File: PD Project/python_code/UART.py
import serial
import struct
import time
import RPi.GPIO as GPIO
from setup import *  # Import all configuration constants

class UART:
    def __init__(self, weight_fill, weight_drain, motor, system, memory, buzzer):
        """Initialize with all required components"""
        self._weight_fill = weight_fill
        self._weight_drain = weight_drain
        self._motor = motor
        self._system = system
        self._memory = memory
        self._buzzer = buzzer
        
        self._buffered = bytearray(4)
        self._buffer_hd = bytearray(BLE_HEADER_LEN)
        self._buffer_data = bytearray(BLE_RX_DATA_LEN)
        self._idx = 0
        self._cmd = 0
        self._len = 0
        self._chksum_rx = 0
        self._chksum_tx = 0
        self._is_streaming = False
        self._is_connected = True
        self._serial = None
        
        # Initialize LED pin
        GPIO.setmode(GPIO.BCM)
        GPIO.setup(PIN_LED, GPIO.OUT)
        GPIO.output(PIN_LED, GPIO.LOW)

    def init(self, baudrate=115200, port='/dev/serial0'):
        """Initialize UART communication"""
        try:
            self._serial = serial.Serial(
                port=port,
                baudrate=baudrate if baudrate else BAUDRATE,
                parity=serial.PARITY_NONE, # no parity bit
                stopbits=serial.STOPBITS_ONE, # one stop bit
                bytesize=serial.EIGHTBITS,
                timeout=0.1  # Short timeout for non-blocking reads
            )
            time.sleep(1)  # Allow time for connection
            
        except Exception as e:
            print(f"UART initialization failed: {e}")
            raise

    def check_connection(self, pin_connect=None):
        """Check Bluetooth connection status"""
        connect_pin = pin_connect 
        currently_connected = GPIO.input(connect_pin)
        
        if currently_connected != self._is_connected:
            if not currently_connected:
                # Connection lost
                self._is_streaming = False
                self._motor.set_state(self._motor.State.STOP.value)
                GPIO.output(PIN_LED, GPIO.LOW)
            
                
            self._is_connected = currently_connected

    def data_buffering(self, val):
        """Buffer and send data with checksum calculation"""
        if isinstance(val, int):
            self._chksum_tx += val
            self._serial.write(bytes([val]))

        elif isinstance(val, float):
            # Pack float into 4 bytes
            packed = struct.pack('f', val)
            for byte in packed:
                self.data_buffering(byte)

    def data_send(self):
        """Send data packet if streaming is enabled"""
        if self._is_streaming and self._serial and self._serial.is_open:
            try:
                # Send packet header
                self.data_buffering(BLE_HEADER)
                self.data_buffering(BLE_HEADER)

                # Prepare data packet
                self._chksum_tx = 0
                self.data_buffering(BLE_TX_DATA_LEN)
                self.data_buffering(self._idx)
                self.data_buffering(self._weight_fill.get_weight())
                self.data_buffering(self._weight_drain.get_weight())
                self.data_buffering(self._system.get_voltage())
                self.data_buffering(self._system.get_charging())
                
                # Send checksum (one's complement)
                self.data_buffering(~self._chksum_tx & 0xFF)
                
                # Increment packet counter
                self._idx = (self._idx + 1) % 100

            except Exception as e:
                print(f"Data send error: {e}")

    def data_receive(self):
        """Receive and process incoming commands"""
        if self._serial and self._serial.in_waiting:
            try:
                # Shift buffer and read new byte
                self.buffer_shift(self._buffer_hd) 
                self._buffer_hd[0] = self._serial.read(1)[0]

                # Check for valid header
                if (self._buffer_hd[3] == BLE_HEADER and 
                    self._buffer_hd[2] == BLE_HEADER):
                    
                    self._cmd = self._buffer_hd[1] #command byte
                    self._len = self._buffer_hd[0] #data length
                    self._chksum_rx = self._cmd + self._len
                    
                    # Read data payload
                    self._buffer_data = self._serial.read(self._len)
                    
                    # Verify checksum
                    for i in range(self._len - 1):
                        self._chksum_rx += self._buffer_data[i]
                    
                    if self._buffer_data[self._len - 1] == (~self._chksum_rx & 0xFF):
                        return self._process_command()
                    
            except Exception as e:
                print(f"Data receive error: {e}")
                
        return False

    def _process_command(self):
        """Execute received command"""
        cmd_char = chr(self._cmd)
        
        if cmd_char == 'R':  # System reset
            self.cleanup()  # Clean up resources first
            subprocess.run(['sudo', 'reboot'])
            
        elif cmd_char == 'D':  # Debug data
            if not self._is_streaming:
                self._print_system_info()
                
        elif cmd_char == 'L':  # LED on
            GPIO.output(PIN_LED, GPIO.HIGH)
        elif cmd_char == 'l':  # LED off
            GPIO.output(PIN_LED, GPIO.LOW)
            
        elif cmd_char == 'S':  # Start streaming
            self._is_streaming = True
            self._motor.set_ready(False)

        elif cmd_char == 's':  # Stop streaming
            self._is_streaming = False
            self._motor.set_ready(True)
            
        elif cmd_char in ['Z', 'z']:  # Tare commands
            self._handle_tare_commands(cmd_char)
            
        elif cmd_char in ['C', 'c']:  # Calibration commands
            self._handle_calibration_commands(cmd_char)
            
        elif cmd_char in ['M', 'm', 'x']:  # Motor commands
            self._handle_motor_commands(cmd_char)
            
        elif cmd_char == 'T':  # Test sequence
            self._run_test_sequence()
            
        elif cmd_char == 'A':  # Alarm command
            p = GPIO.PWM(PIN_BUZZ, 330)  # 330Hz square wave
            p.start(50)                   # 50% duty cycle same as tone
            time.sleep(0.5)               # 500ms duration
            p.stop()    
                              
        elif cmd_char == 'P':  # Play tone
            self._buzzer.beep(5)  # Play musical tone
            
        return True

    def _print_system_info(self):
        """Print system debug information"""
        print(f"\n--- System Info {VERSION} ---")
        print(f"Battery: {self._system.get_voltage():.2f}V")
        print(f"Charging: {'Yes' if self._system.get_charging() >= 200 else 'No'}")
        print(f"Fill Cell: zero={self._weight_fill.get_zero():.0f}, cal={self._weight_fill.get_cal():.2f}")
        print(f"Drain Cell: zero={self._weight_drain.get_zero():.0f}, cal={self._weight_drain.get_cal():.2f}")
        print("----------------------------")

    def _handle_tare_commands(self, cmd):
        """Process tare commands"""
        if cmd == 'Z':
            self._weight_fill.set_zero()
        else:
            self._weight_drain.set_zero()
        self._memory.save()

    def _handle_calibration_commands(self, cmd):
        """Process calibration commands"""
        if cmd == 'C':
            self._weight_fill.set_cal()
        else:
            self._weight_drain.set_cal()
        self._memory.save()

    def _handle_motor_commands(self, cmd):
        """Process motor control commands"""
        if cmd == 'M':
            self._motor.set_state(self._motor.State.UP.value)
        elif cmd == 'm':
            self._motor.set_state(self._motor.State.DOWN.value)
        else:
            self._motor.set_state(self._motor.State.STOP.value)
           

    def _run_test_sequence(self):
        """Execute system test sequence"""
        
        GPIO.output(PIN_BUZZ, GPIO.HIGH)
        self._motor.set_state(self._motor.State.UP.value)
        time.sleep(1)
        self._motor.set_state(self._motor.State.DOWN.value)
        time.sleep(1)
        self._motor.set_state(self._motor.State.STOP.value)
        GPIO.output(PIN_BUZZ, GPIO.LOW)
       

    def buffer_shift(self, buffer):
        """Shift buffer contents by one position"""
        for i in range(len(buffer)-1, 0, -1):
            buffer[i] = buffer[i-1]

    def cleanup(self):
        """Clean up resources"""
        if self._serial and self._serial.is_open:
            self._serial.close()
        GPIO.output(PIN_LED, GPIO.LOW)
       